new/analyse/determine-direct-increment.ts

```diff
@@ -0,0 +1,72 @@
+import type { Package, Message, Module } from '../types';
+import { clone } from 'ramda';
+import { green, red, yellow, white } from 'chalk';
+
+import { logger } from '../helpers/logger';
+
+const { error, log } = logger('determine direct increment');
+
+const mapLevelToIncrementLevel: Record<
+  Message['level'],
+  Package['determinedIncrementLevel']
+> = {
+  misc: -1,
+  patch: 0,
+  minor: 1,
+  major: 2,
+};
+const mapLevelToColour: Record<
+  Message['level'],
+  typeof green | typeof red | typeof yellow
+> = {
+  misc: white,
+  patch: green,
+  minor: yellow,
+  major: red,
+};
+
+export const determineDirectIncrement: Module = (env) => {
+  if (!env.commitMessages) {
+    const errorMessage = '"commitMessage" not present on env config object.';
+    error(errorMessage);
+    throw new Error(errorMessage);
+  }
+  if (!env.packages) {
+    const errorMessage = '"packages" not present on env config object.';
+    error(errorMessage);
+    throw new Error(errorMessage);
+  }
+
+  const clonePackages = clone(env.packages);
+  const packagesToRelease: Record<string, Message['level']> = {};
+
+  log('analyse registered commitMessages');
+  env.commitMessages.forEach((commitMessage) => {
+    commitMessage.affectedPackages?.forEach((affectedPackageName) => {
+      const affectedPackage = clonePackages[affectedPackageName];
+
+      affectedPackage.determinedIncrementLevel = Math.max(
+        affectedPackage.determinedIncrementLevel,
+        mapLevelToIncrementLevel[commitMessage.level]
+      ) as Package['determinedIncrementLevel'];
+
+      // Track affected packages to be releaed with their increemnt level
+      // just for log purposes
+      const possibleLevels = Object.keys(
+        mapLevelToIncrementLevel
+      ) as Message['level'][];
+      packagesToRelease[affectedPackageName] =
```

Comment:
You could directly spit out a message here and avoid a second traversing below.

Reply:
good point

Reply:
oh, wait...
Different commits could be touching one same package with multiple levels (for example: `misc` when updating tests, and `minor` when adding a feature later on).

Is it ok to log increments for the same package more than once?

Reply:
Yep. We're working on the log output anyway in the end.